src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1709,175 +1703,74 @@ public static unsafe nuint WidenAsciiToUtf16(byte* pAsciiBuffer, char* pUtf16Buf
-
-            if (Sse2.IsSupported)
-            {
-                asciiVector = Sse2.LoadVector128(pAsciiBuffer); // unaligned load
-                containsNonAsciiBytes = (uint)Sse2.MoveMask(asciiVector) != 0;
-            }
-            else if (AdvSimd.Arm64.IsSupported)
-            {
-                asciiVector = AdvSimd.LoadVector128(pAsciiBuffer);
-                containsNonAsciiBytes = ContainsNonAsciiByte(asciiVector);
-            }
-            else
-            {
-                throw new PlatformNotSupportedException();
-            }
+            nuint currentOffset = 0;
+            ushort* pCurrentWriteAddress = (ushort*)pUtf16Buffer;
 
-            // If there's non-ASCII data in the first 8 elements of the vector, there's nothing we can do.
+            // Calculating the destination address outside the loop results in significant
+            // perf wins vs. relying on the JIT to fold memory addressing logic into the
+            // write instructions. See: https://github.com/dotnet/runtime/issues/33002
+            nuint finalOffsetWhereCanRunLoop = elementCount - (uint)Vector128<byte>.Count;
 
-            if (containsNonAsciiBytes)
+            do
             {
-                return 0;
-            }
+                Vector128<byte> asciiVector = Vector128.Load(pAsciiBuffer + currentOffset);
 
-            // Then perform an unaligned write of the first part of the input buffer.
+                if (asciiVector.ExtractMostSignificantBits() != 0)
+                {
+                    break;
+                }
 
-            Vector128<byte> zeroVector = Vector128<byte>.Zero;
+                // Vector128.Widen is not used here as it less performant on ARM64
+                Vector128<ushort> utf16HalfVector = Vector128.WidenLower(asciiVector);
+                utf16HalfVector.Store(pCurrentWriteAddress);
+                utf16HalfVector = Vector128.WidenUpper(asciiVector);
+                utf16HalfVector.Store(pCurrentWriteAddress + Vector128<ushort>.Count);
 
-            if (Sse2.IsSupported)
-            {
-                utf16FirstHalfVector = Sse2.UnpackLow(asciiVector, zeroVector);
-                Sse2.Store((byte*)pUtf16Buffer, utf16FirstHalfVector); // unaligned
-            }
-            else if (AdvSimd.IsSupported)
-            {
-                utf16FirstHalfVector = AdvSimd.ZeroExtendWideningLower(asciiVector.GetLower()).AsByte();
-                AdvSimd.Store((byte*)pUtf16Buffer, utf16FirstHalfVector); // unaligned
-            }
-            else
-            {
-                throw new PlatformNotSupportedException();
-            }
+                currentOffset += (nuint)Vector128<byte>.Count;
+                pCurrentWriteAddress += (nuint)Vector128<byte>.Count;
+            } while (currentOffset <= finalOffsetWhereCanRunLoop);
 
-            // Calculate how many elements we wrote in order to get pOutputBuffer to its next alignment
```

Comment:
based on the benchmarking that I've done this was not improving perf in noticeable way, but increasing the code complexity. I've removed it and added `Vector256` code path (less code = less code to duplicate)